src/Pages/ControlPanel.jsx

```diff
@@ -25,24 +25,46 @@ const ControlPanel = () => {
 	const isSmall = useMediaQuery(theme.breakpoints.down("md"));
 
 	const [ctrlVals, dispatchCtrlVals] = useReducer(ctrlValsReducer, { remote: null, local: {} });
-	const fetchRemoteValues = (discardLocal) => {
-		dispatchCtrlVals({ type: "discard_remote" });
-		Promise.all(
-			systemControlsCollections.map(collName => getDocs(collection(db, collName)))
-		).then(querySnapshotArray => {
-			// get an array of arrays of [path, data] pairs, then flatten it to a array of [path, data] pairs
-			const pairs = querySnapshotArray.map(querySnapshot =>
-				// get an array of [path, data] pairs
-				querySnapshot.docs.map(doc => [doc.ref.path, doc.data()])
-			).flat();
-			if (discardLocal) {
-				// do this first to avoid the unnecessary checks in replace_remote
-				dispatchCtrlVals({ type: "discard_local" });
-			}
-			dispatchCtrlVals({ type: "replace_remote", newRemote: Object.fromEntries(pairs) });
-		});
-	};
-	useEffect(() => fetchRemoteValues(), []);
+
+  const fetchRemoteValues = (discardLocal) => {
+    useEffect(() => {
+      dispatchCtrlVals({ type: "discard_remote" });
+
+      
+      //this is the listener for the realtime updates
+      systemControlsCollections.map((collName) => {
+        const collectionRef = query(collection(db, collName));
+        console.log(collName)
+          const unsubscribe = onSnapshot(collectionRef, (querySnapshot) => {
+            const pairs = querySnapshot.docs.map(doc => [doc.ref.path, doc.data()]);
+            if (discardLocal) {
+              // do this first to avoid the unnecessary checks in replace_remote
+              dispatchCtrlVals({ type: "discard_local" });
+            }
+            dispatchCtrlVals({ type: "replace_remote", newRemote: Object.fromEntries(pairs) });
+          });
+          return () => unsubscribe();
```

Comment:
right i normally would unsubscribe() in the useEffect cleanup, but but in this case the function isn't defined in the useEffect scope or so I thought.
Cause there are multiple unsubscribes for each of the systemControlsCollections yk? Why doesn't it work to call unsubscribe() inside the map?

Reply:
Right, it's not defined in the `useEffect` scope. Line 46 is a return statement for the callback used by the map on line 35, right, and it itself returns a callback `unsubscribe` (I'm just gonna write `unsubscribe` since that's equivalent to `() => unsubscribe()`). Simply returning `unsubscribe` would just cause the overarching `systemControlsCollections.map(...)` call to return an array of unsubscribe callbacks. We'd then want to call each of those callbacks when the Effect cleans up, so we need to store them somewhere:
```JS
useEffect(() => {
    let unsubscribes = systemControlsCollections.map(() => {
        /* blah blah */
        return unsubscribe;
    });
    return () => unsubscribes.forEach(unsubscribe => unsubscribe());
})
```
By storing the result of `map` into a variable we retain a handle on the unsubscribe functions which is then visible from the `useEffect` cleanup function.

Reply:
makes lots of sense